### DISCUSSION TITLE
Trying to run the test codes manually

### DISCUSSION THREAD

**Discussion by lakshya-chopra** [open-quantum-safe/oqs-provider#444] [Q&A] [answered]
Hello! I've been trying to compile & execute the test code files provided here - [oqs-provider/test](https://github.com/open-quantum-safe/oqs-provider/tree/main/test) manually using gcc. 
Didn't face any errors in the compilation & execution process, however, I kept encountering this output whenever I tried to run the [oqs_test_tlssig.c](https://github.com/open-quantum-safe/oqs-provider/blob/main/test/oqs_test_tlssig.c) file - **`TLS-SIG handshake test not enabled. Update OpenSSL to more current version.`** I further tried to debug the issue by printing the OpenSSL version at runtime, where I had the latest version (`OpenSSL 3.4.0-dev`). 

I believe that there aren't any issues with my `openssl.cnf` configuration file, since I'm getting **oqs-provider** in the providers list, further I did setup the env vars correctly. 

![image](https://github.com/open-quantum-safe/oqs-provider/assets/77010972/40e62563-6fde-4038-ba32-7ec5a1c0dacc)

<hr/>

Though, there could be some errors with linking, I had used the `-l:oqsprovider.so` & the `-loqs` flags to link the required libraries. 
Here's the entire command I used to compile this file: 
```
gcc tlstest_helpers.o test_common.o oqs_test_tlssig.c -o main -I$WORKSPACE/build/include/openssl -I$WORKSPACE/oqs-provider/oqsprov -L$WORKSPACE/build/lib -lssl -lcrypto -l:oqsprovider.so -loqs
``` 
where the `$WORKSPACE/build/lib` contains the `oqsprovider.so` library alongside others.

Note that all these tests executed perfectly well when using the [scripts/runtests.sh](https://github.com/open-quantum-safe/oqs-provider/blob/main/scripts/runtests.sh) script.

Please let me know if you have any insights or suggestions regarding this. Thank you!

**Comment by baentsch** [comment]
> Note that all these tests executed perfectly well when using the [scripts/runtests.sh](https://github.com/open-quantum-safe/oqs-provider/blob/main/scripts/runtests.sh) script.

If this script works there must be something different in your setup that causes the problems you see. My first guess would be environment variables (whether they're all set correctly in your setup).

What also takes me by surprise is the way you link `oqsprovider` into the executable: Why is `-loqsprovider` not sufficient? 

My recommendation would be for you to scrutinize the build commands executed by the "fullbuild.sh" script (setting `cmake` to verbose mode) and compare it to your manual test build command.

**Reply by lakshya-chopra** [reply]
Thank you for your response. So eventually I was able to run the test & didn't get the same output - `TLS-SIG handshake test not enabled. Update OpenSSL to more current version.` But the issue is that none of signature schemes pass the **TLS-SIG handshake test** and keep giving this error - `SSL routines:ssl_set_cert:unknown certificate type:../ssl/ssl_rsa.c:239:`. 
Furthermore, even the [**scripts/runtests.sh**](https://github.com/open-quantum-safe/oqs-provider/blob/main/scripts/runtests.sh) failed for this specific test (TLS-SIG handshake) this time, despite the other tests running perfectly well. 

For instance, here's **dilithium2** signature scheme being tested:
![image](https://github.com/open-quantum-safe/oqs-provider/assets/77010972/c5c067ac-e9ce-4efb-a491-71846b1c79ae)

And here are the results from the [**runtests.sh**](https://github.com/open-quantum-safe/oqs-provider/blob/main/scripts/runtests.sh) script:
![image](https://github.com/open-quantum-safe/oqs-provider/assets/77010972/8aa408ea-2314-4ac4-b1d6-ba23c0b53ece)

The certificates, keys & the CSRs were generated correctly:
![image](https://github.com/open-quantum-safe/oqs-provider/assets/77010972/6b280e7c-816a-413f-8304-b4704657d39c)

<hr/>

> What also takes me by surprise is the way you link oqsprovider into the executable: Why is -loqsprovider not sufficient?

Well, I'm not really sure about this. I just couldn't get `-loqsprovider` to properly work, maybe because the library's name doesn't start with `lib` (it is `oqsprovider.so`). I renamed it to `liboqsprovider.so` & only then did `-loqsprovider` work. 
[Here's a relevant answer on stackoverflow](https://stackoverflow.com/a/66136692)

**Reply by baentsch** [reply, marked as answer]
> Furthermore, even the [scripts/runtests.sh](https://github.com/open-quantum-safe/oqs-provider/blob/main/scripts/runtests.sh) failed for this specific test (TLS-SIG handshake) this time, despite the other tests running perfectly well.

This would be indicative of an OpenSSL (library) version being used that does not support provider-based signatures (i.e., versions <3.2). Are you 100% certain that you have linked a suitably high version libcrypto and libssl library? What library reference information is `ldd` returning, for example? Is that library >=v3.2?

**Reply by lakshya-chopra** [reply]
Yes, thank you! This was the exact issue, I had two installations of OpenSSL on my device, and the tests were using the older version, which was below `3.2`. So I had to rebuild the `oqs-provider` using the more newer version & then set the `OPENSSL_APP` env to that, and now the tests are working absolutely fine.
